File: processing-service/src/main/java/com/pfm/processing/report/ReportEntry.java
package com.pfm.processing.report;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.pfm.common.domain.NetPosition;
import com.pfm.common.domain.ReportKey;

import java.math.BigDecimal;
import java.time.Instant;
import java.time.LocalDate;
import java.util.Map;

/**
 * One row of the daily summary.
 *
 * <p>The first three components carry the original {@code @JsonProperty} names and
 * are the only ones the CSV uses — they are a frozen contract. Everything after
 * them is additive: decomposed dimensions the UI filters and sorts on, and
 * measures that make a row interpretable (a net of zero is ambiguous without the
 * gross figures).
 */
public record ReportEntry(
        @JsonProperty("Client_Information") String clientInformation,
        @JsonProperty("Product_Information") String productInformation,
        @JsonProperty("Total_Transaction_Amount") long netQuantity,

        String clientType,
        String clientNumber,
        String accountNumber,
        String subaccountNumber,
        String exchangeCode,
        String productGroupCode,
        String symbol,
        LocalDate expirationDate,

        long grossLong,
        long grossShort,
        int tradeCount,
        LocalDate firstTransactionDate,
        LocalDate lastTransactionDate,
        Instant lastUpdatedAt,
        Map<String, BigDecimal> feesByCurrency) {

    public static ReportEntry of(ReportKey key, NetPosition position) {
        return new ReportEntry(
                key.clientInformation(),
                key.productInformation(),
                position.netQuantity(),
                key.clientType(),
                key.clientNumber(),
                key.accountNumber(),
                key.subaccountNumber(),
                key.exchangeCode(),
                key.productGroupCode(),
                key.symbol(),
                key.expirationDate(),
                position.grossLong(),
                position.grossShort(),
                position.tradeCount(),
                position.firstTransactionDate(),
                position.lastTransactionDate(),
                position.lastUpdatedAt(),
                position.feesByCurrency());
    }
}

File: frontend/src/app/report/report-entry.ts
/**
 * One report row. The three PascalCase properties are the frozen legacy contract
 * (they drive the CSV); everything else is additive.
 *
 * Dates arrive as ISO strings ("2010-09-10"), not Date objects — Jackson emits
 * ISO because Spring Boot disables WRITE_DATES_AS_TIMESTAMPS by default.
 */
export interface ReportEntry {
  Client_Information: string;
  Product_Information: string;
  Total_Transaction_Amount: number;

  clientType: string;
  clientNumber: string;
  accountNumber: string;
  subaccountNumber: string;
  exchangeCode: string;
  productGroupCode: string;
  symbol: string;
  expirationDate: string;

  grossLong: number;
  grossShort: number;
  tradeCount: number;
  firstTransactionDate: string | null;
  lastTransactionDate: string | null;
  lastUpdatedAt: string | null;
  /** Keyed by currency code. Values are negative for debits — see the plan's constraints. */
  feesByCurrency: Record<string, number>;
}

/** Provenance of the source file. Run fields are null until an ingestion has happened. */
export interface IngestionStatus {
  configuredPath: string;
  fileExists: boolean;
  fileSizeBytes: number | null;
  fileLastModified: string | null;
  lastIngestAt: string | null;
  fingerprint: string | null;
  totalLines: number | null;
  published: number | null;
  skipped: number | null;
  errorCount: number | null;
}
